core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java

```diff
@@ -117,35 +167,38 @@ public CompletionSuggestionBuilder regex(String regex, RegexOptions regexOptions
      * Note: Only doc values enabled fields are supported
      */
     public CompletionSuggestionBuilder payload(List<String> fields) {
-        this.payloadFields.addAll(fields);
+        this.payloadFields = fields;
         return this;
     }
 
     /**
-     * Sets query contexts for a category context
-     * @param name of the category context to execute on
-     * @param queryContexts a list of {@link CategoryQueryContext}
+     * Sets query contexts for completion
+     * @param queryContexts named query contexts
+     *                      see {@link org.elasticsearch.search.suggest.completion.context.CategoryQueryContext}
+     *                      and {@link org.elasticsearch.search.suggest.completion.context.GeoQueryContext}
      */
-    public CompletionSuggestionBuilder categoryContexts(String name, CategoryQueryContext... queryContexts) {
-        return contexts(name, queryContexts);
-    }
-
-    /**
-     * Sets query contexts for a geo context
-     * @param name of the geo context to execute on
-     * @param queryContexts a list of {@link GeoQueryContext}
-     */
-    public CompletionSuggestionBuilder geoContexts(String name, GeoQueryContext... queryContexts) {
-        return contexts(name, queryContexts);
+    public CompletionSuggestionBuilder contexts(Map<String, List<? extends QueryContext>> queryContexts) {
+        try {
+            XContentBuilder contentBuilder = XContentFactory.jsonBuilder();
+            contentBuilder.startObject();
+            for (Map.Entry<String, List<? extends QueryContext>> contextEntry : queryContexts.entrySet()) {
+                contentBuilder.startArray(contextEntry.getKey());
+                for (ToXContent queryContext : contextEntry.getValue()) {
+                    queryContext.toXContent(contentBuilder, EMPTY_PARAMS);
+                }
+                contentBuilder.endArray();
+            }
+            contentBuilder.endObject();
+            contextBytes = contentBuilder.bytes();
+            return this;
+        } catch (IOException e) {
+            throw new IllegalArgumentException(e);
+        }
     }
 
-    private CompletionSuggestionBuilder contexts(String name, QueryContext... queryContexts) {
-        List<QueryContext> contexts = this.queryContexts.get(name);
-        if (contexts == null) {
-            contexts = new ArrayList<>(2);
-            this.queryContexts.put(name, contexts);
-        }
-        Collections.addAll(contexts, queryContexts);
+    private String field;
```

Comment:
This looks like its only used in innerFromXContent, to circumvent the limitations imposed by ObjectParser and requiring the fieldname at construction time. Can we encapsulate this (and the setter) in an `InnerBuilder extend CompletionSuggestionBuilder` and then use that as the value object the parser fills? Otherwise this extra field feels a bit odd, given that it is not serialized etc... (even though it is private).